File: sdks/ruby/lib/autonoma/graph.rb
# frozen_string_literal: true

require "set"

module Autonoma
  module Graph
    # Topological sort of nodes by FK dependency edges.
    # Returns {"sorted" => [...], "cycles" => [[...]]}.
    def self.topo_sort(nodes, edges)
      node_set = nodes.to_set

      # Filter to only edges between known nodes, skip self-referential
      relevant_edges = edges.select do |e|
        e["from"] != e["to"] && node_set.include?(e["from"]) && node_set.include?(e["to"])
      end

      # Build in-degree map and adjacency list
      in_degree = nodes.each_with_object({}) { |n, h| h[n] = 0 }
      adj = Hash.new { |h, k| h[k] = [] }

      relevant_edges.each do |e|
        # e["from"] depends on e["to"] (from has the FK pointing to to)
        adj[e["to"]] << e["from"]
        in_degree[e["from"]] = (in_degree[e["from"]] || 0) + 1
      end

      # First pass: standard Kahn's
      queue = in_degree.select { |_, d| d == 0 }.keys.sort
      sorted_nodes = []

      until queue.empty?
        node = queue.shift
        sorted_nodes << node
        (adj[node] || []).each do |neighbor|
          in_degree[neighbor] -= 1
          queue << neighbor if in_degree[neighbor] == 0
        end
        queue.sort!
      end

      # Find unsorted nodes
      unsorted = nodes.reject { |n| sorted_nodes.include?(n) }

      return { "sorted" => sorted_nodes, "cycles" => [] } if unsorted.empty?

      cycles = find_sccs(unsorted, relevant_edges)

      return { "sorted" => sorted_nodes, "cycles" => [] } if cycles.empty?

      # Second pass: treat cycle nodes as resolved and re-sort remaining
      cycle_nodes = cycles.flatten.to_set
      still_unsorted = unsorted.reject { |n| cycle_nodes.include?(n) }

      if still_unsorted.any?
        still_set = still_unsorted.to_set
        in_deg2 = still_unsorted.each_with_object({}) { |n, h| h[n] = 0 }
        adj2 = Hash.new { |h, k| h[k] = [] }

        relevant_edges.each do |e|
          if still_set.include?(e["from"]) && still_set.include?(e["to"])
            adj2[e["to"]] << e["from"]
            in_deg2[e["from"]] = (in_deg2[e["from"]] || 0) + 1
          end
        end

        queue2 = in_deg2.select { |_, d| d == 0 }.keys.sort
        until queue2.empty?
          node = queue2.shift
          sorted_nodes << node
          (adj2[node] || []).each do |neighbor|
            in_deg2[neighbor] -= 1
            queue2 << neighbor if in_deg2[neighbor] == 0
          end
          queue2.sort!
        end
      end

      { "sorted" => sorted_nodes, "cycles" => cycles }
    end

    # Find a nullable FK edge in a cycle that can be deferred.
    def self.find_deferrable_edge(cycle, edges)
      cycle_set = cycle.to_set
      edges.find do |e|
        cycle_set.include?(e["from"]) &&
          cycle_set.include?(e["to"]) &&
          e["from"] != e["to"] &&
          e["nullable"] == true
      end
    end

    # Tarjan's SCC algorithm to identify exact cycles among remaining nodes.
    def self.find_sccs(nodes, edges)
      node_set = nodes.to_set
      adj = Hash.new { |h, k| h[k] = [] }

      edges.each do |e|
        adj[e["to"]] << e["from"] if node_set.include?(e["from"]) && node_set.include?(e["to"])
      end

      index_counter = [0]
      stack = []
      on_stack = Set.new
      indices = {}
      lowlinks = {}
      sccs = []

      strong_connect = lambda do |v|
        indices[v] = index_counter[0]
        lowlinks[v] = index_counter[0]
        index_counter[0] += 1
        stack.push(v)
        on_stack.add(v)

        (adj[v] || []).each do |w|
          if !indices.key?(w)
            strong_connect.call(w)
            lowlinks[v] = [lowlinks[v], lowlinks[w]].min
          elsif on_stack.include?(w)
            lowlinks[v] = [lowlinks[v], indices[w]].min
          end
        end

        if lowlinks[v] == indices[v]
          scc = []
          loop do
            w = stack.pop
            on_stack.delete(w)
            scc << w
            break if w == v
          end
          sccs << scc if scc.length > 1
        end
      end

      nodes.each do |node|
        strong_connect.call(node) unless indices.key?(node)
      end

      sccs
    end

    private_class_method :find_sccs
  end
end

File: sdks/go/autonoma/graph.go
package autonoma

import "sort"

// TopoSortResult contains sorted nodes and any detected cycles.
type TopoSortResult struct {
	Sorted []string   `json:"sorted"`
	Cycles [][]string `json:"cycles"`
}

// TopoSort performs topological sort via Kahn's algorithm.
// Returns sorted nodes and any strongly connected components (cycles).
func TopoSort(nodes []string, edges []FKEdge) TopoSortResult {
	nodeSet := make(map[string]bool, len(nodes))
	for _, n := range nodes {
		nodeSet[n] = true
	}

	inDegree := make(map[string]int, len(nodes))
	adj := make(map[string][]string, len(nodes))
	for _, n := range nodes {
		inDegree[n] = 0
		adj[n] = nil
	}

	// Filter to only edges between known nodes, skip self-referential
	var relevantEdges []FKEdge
	for _, e := range edges {
		if e.From != e.To && nodeSet[e.From] && nodeSet[e.To] {
			relevantEdges = append(relevantEdges, e)
		}
	}

	for _, edge := range relevantEdges {
		// edge.From depends on edge.To (From has the FK pointing to To)
		adj[edge.To] = append(adj[edge.To], edge.From)
		inDegree[edge.From]++
	}

	// First pass: standard Kahn's
	var queue []string
	for _, n := range nodes {
		if inDegree[n] == 0 {
			queue = append(queue, n)
		}
	}

	var sorted []string
	for len(queue) > 0 {
		sort.Strings(queue)
		node := queue[0]
		queue = queue[1:]
		sorted = append(sorted, node)
		for _, neighbor := range adj[node] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	// Find unsorted nodes (cycle members + dependents)
	sortedSet := make(map[string]bool, len(sorted))
	for _, s := range sorted {
		sortedSet[s] = true
	}
	var unsortedNodes []string
	for _, n := range nodes {
		if !sortedSet[n] {
			unsortedNodes = append(unsortedNodes, n)
		}
	}

	cycles := findSCCs(unsortedNodes, relevantEdges)
	if len(cycles) == 0 {
		if sorted == nil {
			sorted = []string{}
		}
		return TopoSortResult{Sorted: sorted, Cycles: [][]string{}}
	}

	// Second pass: treat cycle nodes as "resolved" and re-run Kahn's
	cycleNodeSet := make(map[string]bool)
	for _, cycle := range cycles {
		for _, n := range cycle {
			cycleNodeSet[n] = true
		}
	}

	var stillUnsorted []string
	for _, n := range unsortedNodes {
		if !cycleNodeSet[n] {
			stillUnsorted = append(stillUnsorted, n)
		}
	}

	if len(stillUnsorted) > 0 {
		stillSet := make(map[string]bool, len(stillUnsorted))
		for _, n := range stillUnsorted {
			stillSet[n] = true
		}

		inDeg2 := make(map[string]int, len(stillUnsorted))
		adj2 := make(map[string][]string, len(stillUnsorted))
		for _, n := range stillUnsorted {
			inDeg2[n] = 0
			adj2[n] = nil
		}

		for _, edge := range relevantEdges {
			if stillSet[edge.From] && stillSet[edge.To] {
				adj2[edge.To] = append(adj2[edge.To], edge.From)
				inDeg2[edge.From]++
			}
		}

		var queue2 []string
		for _, n := range stillUnsorted {
			if inDeg2[n] == 0 {
				queue2 = append(queue2, n)
			}
		}

		for len(queue2) > 0 {
			sort.Strings(queue2)
			node := queue2[0]
			queue2 = queue2[1:]
			sorted = append(sorted, node)
			for _, neighbor := range adj2[node] {
				inDeg2[neighbor]--
				if inDeg2[neighbor] == 0 {
					queue2 = append(queue2, neighbor)
				}
			}
		}
	}

	if sorted == nil {
		sorted = []string{}
	}

	return TopoSortResult{Sorted: sorted, Cycles: cycles}
}

// findSCCs implements Tarjan's SCC algorithm.
func findSCCs(nodes []string, edges []FKEdge) [][]string {
	if len(nodes) == 0 {
		return nil
	}

	nodeSet := make(map[string]bool, len(nodes))
	for _, n := range nodes {
		nodeSet[n] = true
	}

	adj := make(map[string][]string, len(nodes))
	for _, n := range nodes {
		adj[n] = nil
	}
	for _, edge := range edges {
		if nodeSet[edge.From] && nodeSet[edge.To] {
			adj[edge.To] = append(adj[edge.To], edge.From)
		}
	}

	index := 0
	var stack []string
	onStack := make(map[string]bool)
	indices := make(map[string]int)
	lowlinks := make(map[string]int)
	var sccs [][]string

	var strongConnect func(v string)
	strongConnect = func(v string) {
		indices[v] = index
		lowlinks[v] = index
		index++
		stack = append(stack, v)
		onStack[v] = true

		for _, w := range adj[v] {
			if _, visited := indices[w]; !visited {
				strongConnect(w)
				if lowlinks[w] < lowlinks[v] {
					lowlinks[v] = lowlinks[w]
				}
			} else if onStack[w] {
				if indices[w] < lowlinks[v] {
					lowlinks[v] = indices[w]
				}
			}
		}

		if lowlinks[v] == indices[v] {
			var scc []string
			for {
				w := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				onStack[w] = false
				scc = append(scc, w)
				if w == v {
					break
				}
			}
			if len(scc) > 1 {
				sccs = append(sccs, scc)
			}
		}
	}

	for _, n := range nodes {
		if _, visited := indices[n]; !visited {
			strongConnect(n)
		}
	}

	return sccs
}

// FindDeferrableEdge finds a nullable FK edge in a cycle that can be deferred.
func FindDeferrableEdge(cycle []string, edges []FKEdge) *FKEdge {
	cycleSet := make(map[string]bool, len(cycle))
	for _, n := range cycle {
		cycleSet[n] = true
	}

	for _, edge := range edges {
		if cycleSet[edge.From] && cycleSet[edge.To] && edge.From != edge.To && edge.Nullable {
			e := edge // copy
			return &e
		}
	}
	return nil
}
